### PR TITLE
Histogram Expiry Detector

### PR DESCRIPTION
**Pull request by Uberi**
Ref: https://bugzilla.mozilla.org/show_bug.cgi?id=1149741

Email alerts are sent out listing every histogram that is expired or is soon to be expired. Running it at the moment will result in something like the following:

```
Sending email to rvitillo@mozilla.com with body:
The following histograms will be expired on or before 2015-05-27, and should be removed from the codebase:

* [SUBSCRIBED] BLOCKLIST_SYNC_FILE_LOAD expires in version 35
* NEWTAB_PAGE_SHOWN expires in version 35
* NEWTAB_PAGE_SITE_CLICKED expires in version 35
* TELEMETRY_TEST_EXPIRED expires in version 4.0a1

This is an automated message sent by Cerberus. See https://github.com/mozilla/cerberus for details and source code.
```

### CODE REVIEW

File: alert/expiring.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import json
+import urllib2
+from datetime import datetime, timedelta
+
+from bs4 import BeautifulSoup
+from mail import send_ses
+
+HISTOGRAMS_FILE = "Histograms.json"
+EMAIL_TIME_BEFORE = timedelta(weeks=1)
+FROM_ADDR = "cerberus-alert@mozilla.com"
```
Comment (vitillo) [inline alert/expiring.py]: Please use telemetry-alerts@mozilla.com.

File: alert/expiring.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        version = list(row.find_all("td"))[-1].string.strip().replace("Firefox ", "")
+        version = version.split(".")[0] # only keep the major version
+        date_string = list(row.find_all("th"))[-1].string.strip(" \t\r\n*")
+        try: result[version] = datetime.strptime(date_string, "%Y-%m-%d")
+        except ValueError: pass
+    return result
+
+def is_expiring(histogram_entry, now, release_dates):
+    expiry = histogram_entry.get("expires_in_version", "never").split(".")[0]
+
```
Comment (vitillo) [inline alert/expiring.py]: remove whitespace

File: alert/expiring.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    if expiry not in release_dates: return True
+    
+    # known version, check if it is close to release
+    return release_dates[expiry] - now < EMAIL_TIME_BEFORE
+
+def main():
+    with open(HISTOGRAMS_FILE) as f: histograms = json.load(f)
+    now, release_dates = datetime.now(), get_future_release_dates()
+    notifiable_histograms = sorted([h for h in histograms.items() if is_expiring(h[1], now, release_dates)], key=lambda h: h[0])
+
```
Comment (vitillo) [inline alert/expiring.py]: Please remove whitespace.

File: alert/expiring.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    notifiable_histograms = sorted([h for h in histograms.items() if is_expiring(h[1], now, release_dates)], key=lambda h: h[0])
+    
+    # organize histograms into buckets indexed by email
+    email_histogram_names = {}
+    for name, entry in notifiable_histograms:
+        if entry.get("alert_emails", False):
+            for email in entry["alert_emails"]:
+                if email not in email_histogram_names: email_histogram_names[email] = []
+                email_histogram_names[email].append(name)
+
```
Comment (vitillo) [inline alert/expiring.py]: Please strip the leading whitespace.

File: alert/expiring.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    # check if the version is unknown or far into the future
+    if expiry in {"never", "default"}: return False
+    try:
+        max_release = max(map(int, release_dates.keys()))
+        if int(expiry) > max_release: return False
+    except ValueError: pass
+    if expiry not in release_dates: return True
+    
+    # known version, check if it is close to release
+    return release_dates[expiry] - now < EMAIL_TIME_BEFORE
```
Comment (vitillo) [inline alert/expiring.py]: This takes in consideration only the major version; the version comparison needs to be compatible with our versioning scheme, i.e. versions such as "40.0a1" (v40 nightly) have to be properly supported.

**Comment by vitillo** [comment]
- How has this been tested? Did you use your own AWS credentials?
- Once the comments have been addressed, please add a call to it from run.sh.

**Comment by Uberi** [comment]
Updated:

untitled

File: alert/mail.py [hunk 6]
```diff
@@ -2,11 +2,18 @@
...
 from email.mime.application import MIMEApplication
 from email.mime.multipart import MIMEMultipart
 from email.mime.text import MIMEText
+import json
+
+with open("aws-settings.json", "r") as f:
+  settings = json.load(f)
+  AWS_ACCESS_KEY = settings["AWS_ACCESS_KEY"]
+  AWS_SECRET_ACCESS_KEY = settings["AWS_SECRET_ACCESS_KEY"]
+  AWS_REGION = settings["AWS_REGION"]
```
Comment (vitillo) [inline alert/mail.py]: There is no need for this as boto picks up the credentials from ~/.aws/credentials.

File: alert/mail.py [hunk 7]
```diff
@@ -15,6 +22,13 @@ def send_ses(fromaddr,
...
     """Send an email via the Amazon SES service.
 
+    Configuration is stored in `aws-settings.json` in the working directory, of the following form:
+      {
+        "AWS_ACCESS_KEY": "ACESS_KEY_GOES_HERE",
+        "AWS_SECRET_ACCESS_KEY": "SECRET_ACCESS_KEY_GOES_HERE",
+        "AWS_REGION": "us-west-2"
+      }
+
     Example:
```
Comment (vitillo) [inline alert/mail.py]: Same as above.

File: alert/expiring.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+def is_expiring(histogram_entry, now, release_dates):
+    # check if the expiration version is unknown, invalid, or far into the future
+    expiry = histogram_entry.get("expires_in_version", "never").strip()
+    if expiry in {"never", "default"}: return False
+    try:
+        max_release = next(release_dates.iterkeys())
+        for release in release_dates.iterkeys():
+            if version_compare(max_release, release) < 0:
+                max_release = release
+        if version_compare(expiry, max_release) > 0: return False
```
Comment (vitillo) [inline alert/expiring.py]: This is still incorrect. You are not accounting correctly for the various trees when parsing future branch dates. A histogram might expire for example in nightly (e.g. 40.0a1), in beta (41.0b1), in aurora (e.g. 40.0a2), or in release (41.0), depending on the version number.

**Comment by vitillo** [comment]
In the screenshot you have attached some histograms that expired in earlier releases are shown. This should not happen, only histograms that are about to expire should be shown.

Please add a proper test-suite.

File: alert/expiring.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+EMAIL_TIME_BEFORE = timedelta(weeks=1)
+FROM_ADDR = "telemetry-alert@mozilla.com"
+
+def get_future_release_dates():
+    # scrape for future release date tables
+    response = json.loads(urllib2.urlopen("https://wiki.mozilla.org/api.php?action=parse&format=json&page=RapidRelease/Calendar").read())
+    soup = BeautifulSoup(response["parse"]["text"]["*"])
+    table = soup.find(id="Future_branch_dates").find_parent("h2").find_next_sibling("table")
+    result = {}
+    for i, row in enumerate(table.find_all("tr")):
```
Comment (vitillo) [inline alert/expiring.py]: Please write a comment with the HTML structure you are expecting to find. That will make it easier to maintain in the future.

**Comment by Uberi** [comment]
Updated:

    $ python alert/expiring.py
    Usage: alert/expiring.py list|email|test
      alert/expiring.py list    list histograms that are soon expiring
      alert/expiring.py email   notify users of histograms that are soon expiring
      alert/expiring.py test    run various internal tests

The new expiry detection works as follows:

1. If the release date of the expiry version is known, check if it is expiring soon.
    * This handles cases like those expiring in 40 and 41, when we know exactly when it will be released.
2. Otherwise, if the expiry version is older than all the versions on the release date list, we know the expiry version is in the past, so we ignore it.
    * This handles cases like those expiring in 35 and earlier, when we don't want expired histograms that are too old.
3. Otherwise, find the first version for which the expiry version is before, and check if this version is being released soon.
    * This handles all the other cases, such as those expiring in 41.1b, which will be considered to expire when 42 expires (the next known release date).

**Comment by vitillo** [comment]
That seems overly complicated and I am not sure I understand your reasoning, e.g. a probe set to expire at version 41.0b1 expires when 41.0b1 is released, not when 42.0 is released. Let me clarify what the utility should be doing:

1) Create a map of (version name, release date), see e.g. [1]
2) Translate version names in Histograms.json back to their full version string. As documented in [2] "a version number of type "N" and "N.0" is automatically converted to "N.0a1" in order to expire the histogram also in the development channels. So if a histogram expires in "N" or "N.0", you should map that to "N.0a1". As we might want to expire a probe on a specific channel in the future, you should not assume that all probes are going to expire in Nightly.
3) For each telemetry histogram, retrieve the release date of its expiring version (if *expires_in_version* is set). If that release date is K days (e.g. 2 weeks) in the future, send an alert.

[1] http://en.wikipedia.org/wiki/Firefox_release_history
[2] https://developer.mozilla.org/en-US/docs/Mozilla/Performance/Adding_a_new_Telemetry_probe

**Comment by Uberi** [comment]
Update:

* Histograms are considered soon expiring exactly N days before they actually expire.
* Update tests and docs.
* Parse the channel versions, nightly/aurora/beta/release, out of the release table.

**Comment by vitillo** [comment]
Looks good, thanks!

**Pull request merged**